src/components/Graph.react.js

```diff
@@ -73,208 +29,92 @@ const filterEventData = (gd, eventData, event) => {
+
+        this.state = {
+            extendData: [],
+        };
+
+        this.clearExtendData = this.clearExtendData.bind(this);
     }
 
-    plot(props) {
-        const {figure, animate, animation_options, config} = props;
-        const gd = this.gd.current;
-        if (
-            animate &&
-            this._hasPlotted &&
-            figure.data.length === gd.data.length
-        ) {
-            return Plotly.animate(gd, figure, animation_options);
+    componentDidMount() {
+        if (this.props.extendData) {
+            this.setState({
+                extendData: [this.props.extendData],
+            });
         }
-        return Plotly.react(gd, {
-            data: figure.data,
-            layout: clone(figure.layout),
-            frames: figure.frames,
-            config: config,
-        }).then(() => {
-            const gd = this.gd.current;
-            // double-check gd hasn't been unmounted
-            if (!gd) {
-                return;
-            }
-
-            // in case we've made a new DOM element, transfer events
-            if (this._hasPlotted && gd !== this._prevGd) {
-                if (this._prevGd && this._prevGd.removeAllListeners) {
-                    this._prevGd.removeAllListeners();
-                    Plotly.purge(this._prevGd);
-                }
-                this._hasPlotted = false;
-            }
-
-            if (!this._hasPlotted) {
-                this.bindEvents();
-                Plotly.Plots.resize(gd);
-                this._hasPlotted = true;
-                this._prevGd = gd;
-            }
-        });
     }
 
-    extend(props) {
-        const {extendData} = props;
-        let updateData, traceIndices, maxPoints;
-        if (Array.isArray(extendData) && typeof extendData[0] === 'object') {
-            [updateData, traceIndices, maxPoints] = extendData;
-        } else {
-            updateData = extendData;
-        }
+    componentWillUnmount() {
+        this.setState({
+            extendData: [],
+        });
+    }
 
-        if (!traceIndices) {
-            function getFirstProp(data) {
-                return data[Object.keys(data)[0]];
-            }
+    componentWillReceiveProps(nextProps) {
+        let extendData = this.state.extendData.slice(0);
 
-            function generateIndices(data) {
-                return Array.from(Array(getFirstProp(data).length).keys());
-            }
-            traceIndices = generateIndices(updateData);
+        if (this.props.figure !== nextProps.figure) {
+            extendData = EMPTY_EXTEND_DATA;
         }
 
-        const gd = this.gd.current;
-        return Plotly.extendTraces(gd, updateData, traceIndices, maxPoints);
-    }
+        if (
+            nextProps.extendData &&
+            this.props.extendData !== nextProps.extendData
+        ) {
+            extendData.push(nextProps.extendData);
+        } else {
+            extendData = EMPTY_EXTEND_DATA;
```

Comment:
setting `extendData` to a module constant, then later pushing things into said module constant? Don't we need either a new empty array, or copy into a new longer array?

Reply:
No. Before it's used, the array is copied from the state with `slice`, the state array always being treated as immutable.
https://github.com/plotly/dash-core-components/pull/616/files/efe1e766b4c0ad4b23d04f319649f480d7bf11cf#diff-b2faf9645a4fcca0badb59ef2fba58a3R55

Reply:
Ah got it - thanks.